bpf/lib/lb.h

```diff
@@ -1741,8 +1746,357 @@ lb4_ctx_restore_state(struct __ctx_buff *ctx, struct ct_state *state,
+	__be32	saddr;
+	__be32	daddr;
+	__u8	tos;
+	__wsum csum;
+	int sample_len;
+	int ret;
+	const int inner_offset = sizeof(struct ethhdr) + sizeof(struct iphdr) +
+		sizeof(struct icmphdr);
+
+	if (!revalidate_data(ctx, &data, &data_end, &ip4))
+		return DROP_INVALID;
+
+	/* copy the incoming src and dest IPs and mac addresses to the stack.
+	 * the pointers will not be valid after adding headroom.
+	 */
+
+	if (eth_load_saddr(ctx, smac.addr, 0) < 0)
+		return DROP_INVALID;
+
+	if (eth_load_daddr(ctx, dmac.addr, 0) < 0)
+		return DROP_INVALID;
+
+	saddr = ip4->saddr;
+	daddr = ip4->daddr;
+	tos = ip4->tos;
+
+	/* Resize to ethernet header + 64 bytes or less */
+	sample_len = ctx_full_len(ctx);
+	if (sample_len > ICMP_PACKET_MAX_SAMPLE_SIZE)
+		sample_len = ICMP_PACKET_MAX_SAMPLE_SIZE;
+	ctx_adjust_troom(ctx, sample_len + sizeof(struct ethhdr) - ctx_full_len(ctx));
+
+	data = ctx_data(ctx);
+	data_end = ctx_data_end(ctx);
+
+	/* Calculate the checksum of the ICMP sample */
+	csum = icmp_wsum_accumulate(data + sizeof(struct ethhdr), data_end, sample_len);
+
+	/* We need to insert a IPv4 and ICMP header before the original packet.
+	 * Make that room.
+	 */
+
+#if __ctx_is == __ctx_xdp
+	ret = xdp_adjust_head(ctx, 0 - (int)(sizeof(struct iphdr) + sizeof(struct icmphdr)));
+#else
+	ret = skb_adjust_room(ctx, sizeof(struct iphdr) + sizeof(struct icmphdr),
+			      BPF_ADJ_ROOM_MAC, 0);
+#endif
+
+	if (ret < 0)
+		return DROP_INVALID;
+
+	/* changing size invalidates pointers, so we need to re-fetch them. */
+	data = ctx_data(ctx);
+	data_end = ctx_data_end(ctx);
+
+	/* Bound check all 3 headers at once. */
+	if (data + inner_offset > data_end)
+		return DROP_INVALID;
+
+	/* Write reversed eth header, ready for egress */
+	ethhdr = data;
+	memcpy(ethhdr->h_dest, smac.addr, sizeof(smac.addr));
+	memcpy(ethhdr->h_source, dmac.addr, sizeof(dmac.addr));
+	ethhdr->h_proto = bpf_htons(ETH_P_IP);
+
+	/* Write reversed ip header, ready for egress */
+	ip4 = data + sizeof(struct ethhdr);
+	ip4->version = 4;
+	ip4->ihl = sizeof(struct iphdr) >> 2;
+	ip4->tos = tos;
+	ip4->tot_len = bpf_htons(sizeof(struct iphdr) + sizeof(struct icmphdr) +
+		       (__u16)sample_len);
+	ip4->id = 0;
+	ip4->frag_off = 0;
+	ip4->ttl = 64;
+	ip4->protocol = IPPROTO_ICMP;
+	ip4->check = 0;
+	ip4->daddr = saddr;
+	ip4->saddr = daddr;
+	ip4->check = csum_fold(csum_diff(ip4, 0, ip4, sizeof(struct iphdr), 0));
+
+	/* Write reversed icmp header */
+	icmphdr = data + sizeof(struct ethhdr) + sizeof(struct iphdr);
+	icmphdr->type = ICMP_DEST_UNREACH;
+	icmphdr->code = ICMP_PORT_UNREACH;
+	icmphdr->checksum = 0;
+	icmphdr->un.gateway = 0;
+
+	/* Add ICMP header checksum to sum of its body */
+	csum += csum_diff(icmphdr, 0, icmphdr, sizeof(struct icmphdr), 0);
+	icmphdr->checksum = csum_fold(csum);
+
+	/* Redirect ICMP to the interface we received it on. */
+	cilium_dbg_capture(ctx, DBG_CAPTURE_DELIVERY,
+			   ctx_get_ifindex(ctx));
+	return ctx_redirect(ctx, ctx_get_ifindex(ctx), 0);
+}
```

Comment:
I believe this needs an `edt_set_aggregate(ctx, 0)`, to prevent false-positives in `to-netdev`'s Bandwidth-Manager code.

Reply:
That is a good point, I have not been able to check the interaction with the bandwidth manager. But I suspect you are right. Will have to do a followup.

Reply:
I'm playing around in https://github.com/cilium/cilium/pull/33046.